### ISSUE TITLE
How does one know when headers are finished?

### ISSUE THREAD

**Issue by ioquatix** [igrigorik/http-2#125]
It's possible to have end_headers flag set, but how does one receive that information via https://github.com/igrigorik/http-2/blob/2c127aec0cff711b7f27e9a7f8dad2d952984fd7/lib/http/2/stream.rb#L107 ?

**Comment by HoneyryderChuck** [comment]
I didn't understand the question. The lib emits the headers callback when all headers have been received and decoded.

**Comment by ioquatix** [comment]
How do you know when you've finished emitting headers? i.e. `end_headers`.

**Comment by ioquatix** [comment]
i.e. if there are multiple headers frames.

**Comment by HoneyryderChuck** [comment]
When you're sending, or when you're receiving? 

When it's receiving, the parser receives all the headers/continuation frames, decodes them and emits a single headers callback (haven't tested with trailers though).

When sending, you send all headers at once, or you send them separately, and all calls to #headers before last one will have to set :end_headers to false (as it's done with the #data method).

Does that answer the question?

**Comment by ioquatix** [comment]
Yes, thanks that's a helpful answer.

It wasn't clear to me that there would only be one headers callback.

Once the stream is open, I assume it's not possible to send arbitrary headers frames?

**Comment by ioquatix** [comment]
Okay so I checked the code and it looks like you and send multiple frames of headers down the wire.

https://github.com/igrigorik/http-2/blob/2c127aec0cff711b7f27e9a7f8dad2d952984fd7/lib/http/2/stream.rb#L151-L157

Where is it reassembled?

**Comment by HoneyryderChuck** [comment]
You definitely can send. And they'll be reassembled at the server, according to protocol.

**Comment by ioquatix** [comment]
Okay, I may investigate this further, because it looks like there is a potential issue here.

**Comment by ioquatix** [comment]
So, if one wants to send multiple frames of headers, do you need to manually use a `HEADERS` followed by a `CONTINUATION` frame?

**Comment by ioquatix** [comment]
Is it actually valid to send multiple `HEADERS` frames?

**Comment by HoneyryderChuck** [comment]
You don't need to manually send frames, the lib handles that for you. You can however call #headers multiple times, provided that you'll set :end_headers as false in all the before-last calls. Internally, it will create all the necessary frames.

**Comment by ioquatix** [comment]
Okay, so there will only ever be one `:headers` event emitted and it will be complete. After that, we can expect 0 or more data frames. Let me know if this is wrong.

**Comment by ioquatix** [comment]
I did a bit more digging into this after experiencing issues with Cloudflare and a few other situations.

I found it's feasible to send multiple header frames, but `end_headers` is not preserved correctly on the receiver end in the case of this code base.

It's logical to only have one headers frame followed by continuation frames if required.

That being said, this library doesn't enforce any of those invariants and can break if you try to do so.
